libsolidity/formal/BMC.cpp

```diff
@@ -436,12 +450,29 @@ bool BMC::visit(ForStatement const& _node)
 		forConditionOnPreviousIterations = forConditionOnPreviousIterations && forCondition;
 	}
 	if (bmcLoopIterations > 0)
-		m_context.addAssertion(not(forCondition) || broke);
+	{
+		//after loop iterations are done, we check the loop condition last final time
+		auto indices = copyVariableIndices();
+		if (_node.condition())
+		{
+			_node.condition()->accept(*this);
+			forCondition = expr(*_node.condition());
+		}
+		// asseert that the loop is complete
+		m_context.addAssertion(!forCondition || broke || !forConditionOnPreviousIterations);
+		mergeVariables(
+			broke || !forConditionOnPreviousIterations,
+			indices,
+			copyVariableIndices()
+		);
```

Comment:
Are you sure we need this `mergeVariables` here, instead of stopping at `addAssertion`? I don't really understand why this is needed

Reply:
We have to merge it. It may happen that accepting the condition changes state of some variables and we don't want that in case loop has already been completed.

Reply:
yea good point. I think do-while doesn't have the merge though, does it?

Reply:
yes, do while doesn't need that